lib/spack/spack/solver/concretize.lp

```diff
@@ -689,15 +689,29 @@ requirement_group_satisfied(node(ID, Package), X) :-
 % flags if their only source is from a requirement. This is overly-specific
 % and should use a more-generic approach like in https://github.com/spack/spack/pull/37180
 
-{ attr("node_flag", node(ID, A1), A2, A3) } :-
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_set", A1, A2, A3).
-
-{ attr("node_flag_source", node(ID, A1), A2, node(ID, A3)) } :-
-  requirement_group_member(Y, Package, X),
-  activate_requirement(node(ID, Package), X),
-  imposed_constraint(Y,"node_flag_source", A1, A2, A3).
+{ attr("node_flag", node(ID, SubPackage), FlagType, FlagValue) } :-
+  requirement_group_member(ConditionID, SubPackage, RequirementID),
+  activate_requirement(node(ID, SubPackage), RequirementID),
+  pkg_fact(SubPackage, condition_effect(ConditionID, EffectID)),
+  imposed_constraint(EffectID, "node_flag_set", SubPackage, FlagType, FlagValue).
+
+{ attr("node_flag", node(SubPackageID, SubPackage), FlagType, FlagValue) } :-
+  requirement_group_member(ConditionID, Package, RequirementID),
+  activate_requirement(node(PackageID, Package), RequirementID),
+  pkg_fact(Package, condition_effect(ConditionID, EffectID)),
+  imposed_constraint(EffectID, "node_flag_set", Package, FlagType, FlagValue),
+  provider(node(SubPackageID, SubPackage), node(PackageID, Package)).
```

Comment:
@alalazo

Based on  https://github.com/spack/spack/pull/39880#discussion_r1328593642 I added this, although the updated test doesn't pass (i.e. this is not enforced).

I added the below constraint, since it seemed like the package wasn't being marked as a provider by the concretizer:

```
1 { attr("virtual_node", node(0..X-1, Virtual)) : max_dupes(Virtual, X) } :- 
  virtual_condition_holds(node(ProviderID, Provider), Virtual),
  attr("node", node(ProviderID, Provider)).
```

it turns out with that constraint, you can't do `spack spec openmpi`, which seems wrong to me (my statement is "if there is a node and it meets the condition to provide a virtual, then there should be a virtual node for that").

I'm inclined to do one of:

* Drop support for cflags on requirements associated with virtuals (only allow it for non-virtuals)
* or, use a generic approach like https://github.com/spack/spack/pull/37180

Reply:
> it turns out with that constraint, you can't do `spack spec openmpi`

I should further clarify that this also fails when adding just this clause to the `concretize.lp` on develop (i.e. including none of the other constraints here).

Reply:
I'll mark this as resolved since other updates have shed some light on why this wasn't working.